spring-cloud-azure-feature-management/src/main/java/com/microsoft/azure/spring/cloud/feature/manager/feature/filters/TargetingFilter.java

```diff
@@ -0,0 +1,191 @@
+            for (String group : targetingContext.getGroups()) {
+                Optional<GroupRollout> groupRollout = settings.getAudience().getGroups().stream()
+                        .filter(g -> compairStrings(g.getName(), group)).findFirst();
+
+                if (groupRollout.isPresent()) {
+                    String audienceContextId = targetingContext.getUserId() + "\n" + context.getName() + "\n" + group;
+
+                    if (isTargeted(audienceContextId, groupRollout.get().getRolloutPercentage())) {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        String defaultContextId = targetingContext.getUserId() + "\n" + context.getFeatureName();
+
+        return isTargeted(defaultContextId, settings.getAudience().getDefaultRolloutPercentage());
+    }
+
+    private boolean isTargeted(String contextId, double percentage) {
+        byte[] hash = null;
+
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            hash = digest.digest(contextId.getBytes());
+        } catch (NoSuchAlgorithmException e) {
+            throw new TargetingException("Unable to find SHA-256 for targeting.", e);
+        }
+
+        if (hash == null) {
+            throw new TargetingException("Unable to create Targeting Hash for " + contextId);
+        }
+        
+        ByteBuffer wrapped = ByteBuffer.wrap(hash);
+        int contextMarker = Math.abs(wrapped.getInt());
+
+        double contextPercentage = (contextMarker / (double) Integer.MAX_VALUE) * 100;
+        return contextPercentage < percentage;
+    }
+
+    private boolean tryValidateSettings(TargetingFilterSettings settings) {
+        String paramName = "";
+        String reason = "";
+
+        if (settings.getAudience() == null) {
+            paramName = AUDIENCE;
+            reason = REQUIRED_PARAMETER;
+            
+            throw new TargetingException(paramName + " : " + reason);
+        }
+
+        Audience audience = settings.getAudience();
+        if (audience.getDefaultRolloutPercentage() < 0
+                || audience.getDefaultRolloutPercentage() > 100) {
+            paramName = AUDIENCE + "." + audience.getDefaultRolloutPercentage();
+            reason = OUT_OF_RANGE;
+
+            throw new TargetingException(paramName + " : " + reason);
+        }
+
+        List<GroupRollout> groups = audience.getGroups();
+        if (groups != null) {
+            int index = 0;
+
+            for (GroupRollout groupRollout : groups) {
+                if (groupRollout.getRolloutPercentage() < 0 || groupRollout.getRolloutPercentage() > 100) {
+                    paramName = AUDIENCE + "[" + index + "]." + groups.get(index).getRolloutPercentage();
+                    reason = OUT_OF_RANGE;
+
+                    throw new TargetingException(paramName + " : " + reason);
+                }
+                index++;
+            }
```

Comment:
Given that the index is needed, what do you think about using regular for with index?

```java
for (int index = 0; index < groups.size(); index++) {
                GroupRollout groupRollout = groups.get(index);
                if (groupRollout.getRolloutPercentage() < 0 || groupRollout.getRolloutPercentage() > 100) {
                    paramName = AUDIENCE + "[" + index + "]." + groups.get(index).getRolloutPercentage();
                    reason = OUT_OF_RANGE;

                    throw new TargetingException(paramName + " : " + reason);
                }
            }
```